File: src/main/java/com/chatchatabc/jpademojava/domain/service/UserService.java
package com.chatchatabc.jpademojava.domain.service;

import com.chatchatabc.jpademojava.domain.model.User;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public interface UserService extends UserDetailsService {

    /**
     * Register new user
     *
     * @param user
     * @return
     */
    User register(User user);

    /**
     * Update user profile
     *
     * @param userId
     * @param newUserInfo
     * @return
     */
    @Transactional
    User update(String userId, User newUserInfo) throws Exception;

    /**
     * Update user password
     *
     * @param userId
     * @param oldPassword
     * @param newPassword
     * @return
     */
    @Transactional
    User updatePassword(String userId, String oldPassword, String newPassword) throws Exception;
}

File: src/main/kotlin/com/chatchatabc/jpademo/domain/service/UserService.kt
package com.chatchatabc.jpademo.domain.service

import com.chatchatabc.jpademo.application.dto.user.UserProfileUpdateRequest
import com.chatchatabc.jpademo.application.dto.user.UserRegisterRequest
import com.chatchatabc.jpademo.domain.model.User
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
interface UserService : UserDetailsService {

    /**
     * User Register
     */
    fun register(user: UserRegisterRequest): User

    /**
     * Update User Profile
     */
    @Transactional
    fun update(userId: String, user: UserProfileUpdateRequest): User

    /**
     * Update User Password
     */
    @Transactional
    fun updatePassword(userId: String, oldPassword: String, newPassword: String): User
}